grails-app/services/au/org/ala/userdetails/PasswordService.groovy

```diff
@@ -29,32 +35,237 @@ class PasswordService {
+    }
+
+    /**
+     * Validate a password against a password policy.
+     * @param username
+     * @param password
+     * @return
+     */
+    Map<String, Object> validatePassword(String username, String password) {
+        // policy is:
+        // - length of 8 - 64 characters
+        // - at least 3 of these:
+        //      - upper case letter, e.g. A, B, C
+        //      - lower case e.g. a, b, c
+        //      - numerals e.g. 1, 2, 55
+        //      - non-alphanumeric e.g. $, *, #, &
+        // - must not be a commonly used password (which also accounts for some passwords that are entirely dictionary words)
+        // - must not include the part of the email before the '@', as email addresses are used as usernames
+        // - must not contain a sequence of 5 or more letters that appear next to each other on a US stand keyboard
+
+        def passwordData = new PasswordData(username, password, PasswordData.Origin.User)
+
+        // reference: https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls
+        def ruleLength = new LengthRule(8, 64)
+
+        // ensure password does not contain the user name
+        def ruleUsername = new UsernameRule(false, true, MatchBehavior.Contains)
+
+        // don't allow sequences of characters that are next to each other on a standard US qwerty keyboard
+        def ruleSequence = new IllegalSequenceRule(EnglishSequenceData.USQwerty)
+
+        // at least x of y
+        def ruleGroup = new CharacterCharacteristicsRule()
+
+        // at least 3 of the rules must be met
+        ruleGroup.setNumberOfCharacteristics(3)
+
+        // The rules that need to be met
+        ruleGroup.getRules().addAll(getPasswordCharacterRules())
+
+        // password must not be in the list of common passwords
+        // source https://raw.githubusercontent.com/danielmiessler/SecLists/5e1dc9cc79aac54b373349e2a97bbb22f1b63bb3/Passwords/Common-Credentials/10-million-password-list-top-100000.txt
+        def commonPassInputStream = this.class.classLoader.getResourceAsStream("common-passwords.txt")
+        def commonPassInputStreamReaders = [new InputStreamReader(commonPassInputStream)] as Reader[]
+        def caseSensitive = false
+        def sorting = new ArraysSort()
+        def commonPasswordsList = WordLists.createFromReader(commonPassInputStreamReaders, caseSensitive, sorting)
+        def commonPasswordsDict = new WordListDictionary(commonPasswordsList)
+        def ruleCommonPasswords = new DictionaryRule(commonPasswordsDict)
+
+        def validator = new PasswordValidator(ruleLength, ruleUsername, ruleSequence, ruleGroup, ruleCommonPasswords)
+
+        // at least 20 bits of entropy
+        // not enforcing this because this measures the entropy of the password generation process, which we don't control
+        // reference https://csrc.nist.gov/publications/detail/sp/800-63b/final
+        double entropy
```

Comment:
It seems like `entropy` in the result isn't used outside of tests and could easily be removed